<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="TUSCAN">
<title>Introduction to TUSCAN • TUSCAN</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to TUSCAN">
<meta property="og:description" content="TUSCAN">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TUSCAN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Intro_to_TUSCAN.html">Introduction to TUSCAN</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to TUSCAN</h1>
                        <h4 data-toc-skip class="author">Chenxuan Zang,
Charles C. Guo, Peng Wei, and Ziyi Li</h4>
            
      
      
      <div class="d-none name"><code>Intro_to_TUSCAN.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TUSCAN</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/magick/" class="external-link">magick</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="loading-required-input-data">Loading required input data<a class="anchor" aria-label="anchor" href="#loading-required-input-data"></a>
</h2>
<p>TUSCAN requires three types of input data:</p>
<ul>
<li>Spatial transcriptomics count data.<br>
</li>
<li>Spatial location information.<br>
</li>
<li>Original histology image.</li>
</ul>
<p>We used the HER2-positive breast cancer sample data (patient B,
replicate 1) from the 2021 Andersson’s paper for the illustration in
this vignette. The gene expression profile was measured from the Spatial
Transcriptomics (ST) platform, with 293 spots and 16,148 genes
detected.</p>
<div class="section level3">
<h3 id="spatial-transcriptomics-count-data">Spatial transcriptomics count data<a class="anchor" aria-label="anchor" href="#spatial-transcriptomics-count-data"></a>
</h3>
<p>The spatial raw count data should be a <span class="math inline">\(N
\times D\)</span> matrix, with rows (<span class="math inline">\(N\)</span>) representing genes and columns (<span class="math inline">\(D\)</span>) representing spots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the example spatial count data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_count"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">example_count</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;            10x13 10x14 10x15 10x16</span></span>
<span><span class="co">#&gt; FO538757.1     0     0     0     1</span></span>
<span><span class="co">#&gt; SAMD11         0     0     0     0</span></span>
<span><span class="co">#&gt; NOC2L          0     0     0     0</span></span>
<span><span class="co">#&gt; KLHL17         0     0     0     0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatial-location-information">Spatial location information<a class="anchor" aria-label="anchor" href="#spatial-location-information"></a>
</h3>
<p>The spatial location data should be a dataframe where each row
represents a spatial spot. This dataframe contains two columns ‘x’ and
‘y’, with ‘x’ representing the row pixel coordinate of the spot on the
histology image and ‘y’ representing the column pixel coordinate of the
spot on the histology image. The rownames of the spatial location
dataframe should match exactly with the colnames of the spatial count
data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the example spatial location data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_loc"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">example_loc</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;             x       y</span></span>
<span><span class="co">#&gt; 10x13 2573.12 3493.69</span></span>
<span><span class="co">#&gt; 10x14 2559.81 3790.17</span></span>
<span><span class="co">#&gt; 10x15 2574.28 4075.00</span></span>
<span><span class="co">#&gt; 10x16 2575.73 4373.23</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="histology-image">Histology image<a class="anchor" aria-label="anchor" href="#histology-image"></a>
</h3>
<p>The input image should be the original high-resolution H&amp;E image
of the measured sample.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the example histology image</span></span>
<span><span class="va">image_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"example.jpg"</span>, package <span class="op">=</span> <span class="st">"TUSCAN"</span><span class="op">)</span></span>
<span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="va">image_path</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="Intro_to_TUSCAN_files/figure-html/unnamed-chunk-4-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>Here provide an example code to read in 10x Visium data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">count.data</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"path/to/name_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">loc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"path/to/tissue_positions_list.csv"</span><span class="op">)</span></span>
<span><span class="va">loc.data</span> <span class="op">&lt;-</span> <span class="va">loc.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span> <span class="op">(</span><span class="va">X0</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">loc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">loc.data</span><span class="op">[</span>,<span class="fl">6</span><span class="op">]</span>,y<span class="op">=</span><span class="va">loc.data</span><span class="op">[</span>,<span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">image_path</span> <span class="op">&lt;-</span> <span class="st">"path/to/image"</span></span>
<span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="va">image_path</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-tuscan-object">Create TUSCAN object<a class="anchor" aria-label="anchor" href="#create-tuscan-object"></a>
</h3>
<p>We can use <code><a href="../reference/createTUSCANObject.html">createTUSCANObject()</a></code> to create a TUSCAN
object. The required inputs are raw count matrix, location, and image.
<code>project</code> is a parameter that allows users to name their own
project. <code>r</code> is an image smoothing parameter. A detailed
explanation could refer to <code><a href="../reference/createTUSCANObject.html">help(createTUSCANObject)</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createTUSCANObject.html">createTUSCANObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">example_count</span>, </span>
<span>                          location <span class="op">=</span> <span class="va">example_loc</span>, </span>
<span>                          img <span class="op">=</span> <span class="va">img</span>, </span>
<span>                          project<span class="op">=</span><span class="st">"TUSCAN"</span>, </span>
<span>                          r<span class="op">=</span><span class="fl">49</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="identify-normal-reference">Identify normal reference<a class="anchor" aria-label="anchor" href="#identify-normal-reference"></a>
</h2>
<div class="section level3">
<h3 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h3>
<p>We first perform dimensional reduction on the spatial gene count
data. This step is realized by employing an autoencoder on the top
<code>nhvg</code> (default is 2000) highly variable genes (HVGs).
<code>d</code> is the number of dimensions after reduction (default is
20).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reduceDim.html">reduceDim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, nhvg<span class="op">=</span><span class="fl">2000</span>,d<span class="op">=</span><span class="fl">20</span>,epochs<span class="op">=</span><span class="fl">50</span>,batch_size<span class="op">=</span><span class="fl">64</span>, seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Start dimension reduction by autoencoder..."</span></span>
<span><span class="co">#&gt; 10/10 - 0s - 327ms/epoch - 33ms/step</span></span>
<span><span class="co">#&gt; [1] "Dimension reduction finished..."</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preliminary-clustering">Preliminary clustering<a class="anchor" aria-label="anchor" href="#preliminary-clustering"></a>
</h3>
<p>To accurately infer copy number variations (CNV) through SRT data, we
need to first recognize a subset of normal spots as the reference
baseline values. So we first conduct a preliminary clustering based on
the spatial gene expression data and histology image. Since more than
one cluster may comprise normal cells, we recommend over clustering at
this step to identify the most likely cluster with the highest purity as
the normal cluster.</p>
<p><code><a href="../reference/PreCluster.html">PreCluster()</a></code> functions clusters the spots. It provides
two clustering methods:<code>km</code>(K-Means) and <code>louvain</code>
(Louvain). For K-Means method, the number of clusters <code>k</code>
needs to be specified. For louvain clustering, the number of clusters
can be adjusted by the parameter <code>res</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pre-clustering</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PreCluster.html">PreCluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, method <span class="op">=</span> <span class="st">'km'</span>, k<span class="op">=</span><span class="fl">6</span>, seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Found 6 clusters"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-spatial-clusters">Visualizing spatial clusters<a class="anchor" aria-label="anchor" href="#visualizing-spatial-clusters"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">obj</span>, class <span class="op">=</span> <span class="st">"cluster"</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Intro_to_TUSCAN_files/figure-html/unnamed-chunk-8-1.png" width="90%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="determine-the-cluster-selected-as-the-reference">Determine the cluster selected as the reference<a class="anchor" aria-label="anchor" href="#determine-the-cluster-selected-as-the-reference"></a>
</h3>
<p><code><a href="../reference/FindNormalCluster.html">FindNormalCluster()</a></code> function can print the name of group
from the previous clustering result that is most probable to be the
normal reference.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FindNormalCluster.html">FindNormalCluster</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 40 </span></span>
<span><span class="co">#&gt; number of iterations= 59 </span></span>
<span><span class="co">#&gt; number of iterations= 65 </span></span>
<span><span class="co">#&gt; number of iterations= 63 </span></span>
<span><span class="co">#&gt; number of iterations= 66 </span></span>
<span><span class="co">#&gt; number of iterations= 137 </span></span>
<span><span class="co">#&gt; [1] "Cluster 1 is selected as the normal reference."</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="inferring-the-spatial-copy-number-profile">Inferring the spatial copy number profile<a class="anchor" aria-label="anchor" href="#inferring-the-spatial-copy-number-profile"></a>
</h2>
<p>Next we can run the <code><a href="../reference/runCNV.html">runCNV()</a></code> function to construct the
CNV profile. The essential parameters are:</p>
<ul>
<li>
<code>object</code>: TUSCAN object.<br>
</li>
<li>
<code>ref_group_names</code>: A vector containing the cluster
name(s) of the reference spots used for inferring CNV.<br>
</li>
<li>
<code>min_ave_counts_per_gene</code>: Minimum average gene
expression level across all spots (default: 0.01).<br>
</li>
<li>
<code>min_spots_per_gene</code>: Minimum number of spots requiring
expression measurements to include the corresponding gene (default:
3).<br>
</li>
<li>
<code>chr_exclude</code>: List of chromosomes in the reference
genome annotations that should be excluded from analysis (default:
c(‘chrX’, ‘chrY’, ‘chrM’).<br>
</li>
<li>
<code>normalization</code>: Whether perform normalization step. We
suggest to skip this step if the SRT data is generated from ST platform
(default: TRUE).</li>
</ul>
<p>The computed copy number matrix is stored in the slot
<code>cnv.data</code> of the TUSCAN object. More details of
<code><a href="../reference/runCNV.html">runCNV()</a></code> function could refer to
<code><a href="../reference/runCNV.html">help(runCNV)</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnv_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCNV.html">runCNV</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, </span>
<span>                  ref_group_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">)</span>,</span>
<span>                  min_ave_counts_per_gene<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>                  min_spots_per_gene<span class="op">=</span><span class="fl">3</span>,</span>
<span>                  chr_exclude<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chrX'</span>, <span class="st">'chrY'</span>, <span class="st">'chrM'</span><span class="op">)</span>,</span>
<span>                  normalization <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "create pre-CNV analysis object ..."</span></span>
<span><span class="co">#&gt; [1] "filtering data ..."</span></span>
<span><span class="co">#&gt; [1] "15109 genes and 295 spots in raw count matrix"</span></span>
<span><span class="co">#&gt; [1] "filtered out 2357 genes with less than 0.01 average counts; remaining 12752 genes"</span></span>
<span><span class="co">#&gt; [1] "filtered out 70 genes that expressed in less than 3 spots; remaining 12682 genes"</span></span>
<span><span class="co">#&gt; [1] "find 12092 mathed genes in the count matrix and gene order file"</span></span>
<span><span class="co">#&gt; [1] "start cnv analysis process ..."</span></span>
<span><span class="co">#&gt; [1] "step 1: perform log transformation ..."</span></span>
<span><span class="co">#&gt; [1] "step 2: subtract average normal spots gene expression ..."</span></span>
<span><span class="co">#&gt; [1] "step 3: Apply maximum centered expression thresholds to data ..."</span></span>
<span><span class="co">#&gt; [1] "step 4: smooth the data along chromosome with gene windows ..."</span></span>
<span><span class="co">#&gt; [1] "step 5: center spots/observations after smoothing ..."</span></span>
<span><span class="co">#&gt; [1] "step 6: subtract average reference ..."</span></span>
<span><span class="co">#&gt; [1] "step 7: invert log transform ..."</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualizing-spatial-cnv-by-heatmap">Visualizing spatial CNV by heatmap<a class="anchor" aria-label="anchor" href="#visualizing-spatial-cnv-by-heatmap"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCNVheatmap.html">plotCNVheatmap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cnv_obj</span>, title <span class="op">=</span> <span class="st">"heatmap"</span>, output_format <span class="op">=</span> <span class="st">"png"</span><span class="op">)</span></span>
<span><span class="co">#&gt; agg_png </span></span>
<span><span class="co">#&gt;       2</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">image</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="st">"heatmap.png"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">image</span><span class="op">)</span></span></code></pre></div>
<p><img src="Intro_to_TUSCAN_files/figure-html/unnamed-chunk-12-1.png" width="90%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="identify-tumor-region">Identify tumor region<a class="anchor" aria-label="anchor" href="#identify-tumor-region"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TumorCluster.html">TumorCluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cnv_obj</span>, seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualizing-tumor-region">Visualizing tumor region<a class="anchor" aria-label="anchor" href="#visualizing-tumor-region"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">tumor</span>, class <span class="op">=</span> <span class="st">"tumor"</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Intro_to_TUSCAN_files/figure-html/unnamed-chunk-14-1.png" width="90%" style="display: block; margin: auto;"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chenxuan Zang, Peng Wei, Ziyi Li, Charles C. Guo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
