<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs data filtering and whole genome copy number inferring."><title>Inferring copy number profile (CNP). — runCNV • TUSCAN</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Inferring copy number profile (CNP). — runCNV"><meta property="og:description" content="This function performs data filtering and whole genome copy number inferring."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">TUSCAN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Intro_to_TUSCAN.html">Introduction to TUSCAN</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Inferring copy number profile (CNP).</h1>
      
      <div class="d-none name"><code>runCNV.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs data filtering and whole genome copy number inferring.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runCNV</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  gene_order_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotations_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">ref_group_names</span>,</span>
<span>  min_ave_counts_per_gene <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  min_spots_per_gene <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  chr_exclude <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrX"</span>, <span class="st">"chrY"</span>, <span class="st">"chrM"</span><span class="op">)</span>,</span>
<span>  max_centered_threshold <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  normalization <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalize_factor <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  window_length <span class="op">=</span> <span class="fl">101</span>,</span>
<span>  denoise <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>TUSCAN object.</p></dd>


<dt>gene_order_file</dt>
<dd><p>Gene location file, containing chromosome, start, and stop position for each gene.</p>
<p>example:
chr  start   stop
DDX11L1 chr1  11869  14412
WASH7P  chr1  14363  29806
FAM138A chr1  34554  36081
OR4F5   chr1  69091  70008
OR4F29  chr1 367640 368634
OR4F16  chr1 621059 622053
...</p></dd>


<dt>annotations_file</dt>
<dd><p>File containing annotations of all spots, with two columns representing the spot barcode and the spot clustering name (support '.txt' and '.csv' file).
If the user do not provide an annotation file, the <code>pre_cluster</code> result will be used as the annotations.</p>
<p>example:
V1   V2
10x13    1
10x14    2
10x15    1
10x16    1
...</p></dd>


<dt>ref_group_names</dt>
<dd><p>A vector containing the cluster name of the reference (normal) spots used for inferring CNV.</p></dd>


<dt>min_ave_counts_per_gene</dt>
<dd><p>Minimum average gene expression level across all spots (default: 0.01).</p></dd>


<dt>min_spots_per_gene</dt>
<dd><p>Minimum number of spots requiring expression measurements to include the corresponding gene (default: 3).</p></dd>


<dt>chr_exclude</dt>
<dd><p>List of chromosomes in the reference genome annotations that should be excluded from analysis (default: c('chrX', 'chrY', 'chrM').</p></dd>


<dt>max_centered_threshold</dt>
<dd><p>Maximum value a value can have after centering. Also sets a lower bound of -1 * this value (default: 3).</p></dd>


<dt>normalization</dt>
<dd><p>Whether perform normalization step. We suggest to skip this step if the SRT data is generated from ST platform (default: TRUE).</p></dd>


<dt>normalize_factor</dt>
<dd><p>If the normalize factor is null (default setting), normalize_factor=median(UMI per spot)/UMI per spot.</p></dd>


<dt>window_length</dt>
<dd><p>Length of the window for the moving average (smoothing). Should be an odd integer (default: 101).</p></dd>


<dt>denoise</dt>
<dd><p>Whether perform de-noise step (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Return TUSCAN object containing inferred CNV data</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chenxuan Zang, Peng Wei, Ziyi Li, Charles C. Guo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

